source_middleware.go

```diff
@@ -34,3 +54,283 @@ func SourceWithMiddleware(d Source, middleware ...SourceMiddleware) Source {
 	}
 	return d
 }
+
+// -- SourceWithSchema ----------------------------------------------
+
+const (
+	configSourceSchemaFormat         = "sdk.schema.format"
+	configSourceSchemaPayloadEncode  = "sdk.schema.payload.encode"
+	configSourceSchemaPayloadSubject = "sdk.schema.payload.subject"
+	configSourceSchemaKeyEncode      = "sdk.schema.key.encode"
+	configSourceSchemaKeySubject     = "sdk.schema.key.subject"
+)
+
+// SourceWithSchema is a middleware that extracts and encodes the record payload
+// and key with a schema. The schema is extracted from the record data for each
+// record produced by the source. The schema is registered with the schema
+// service and the schema subject is attached to the record metadata.
+type SourceWithSchema struct {
```

Comment:
Nit: maybe name the middleware so that it reflects that the schema is automatically extracted/registered? The middleware is doing a lot, which makes it a bit more difficult to find a good name.:)

Reply:
That's a valid concern, I was also thinking about what to call this middleware but didn't spend too much time figuring out a better suitable name. Maybe `SourceWithAutoSchemaExtraction`? 🤔

Reply:
To me at least, `SourceWithSchemaExtraction` sounds good enough, it kinda hints that the extraction is being done automatically.